<a href="https://colab.research.google.com/github/nurfnick/Mathematical_Musings/blob/main/GameTheory.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install nashpy

  Created wheel for nashpy: filename=nashpy-0.0.21-py3-none-any.whl size=15281 sha256=69e18713054c02cb0673a01469649557cfb1ada536ebeb7fbf8537864355b567
  Stored in directory: /root/.cache/pip/wheels/02/08/62/cf4fa931e0a317d180936b266169a57f4bb4eb801465bbe8a1
Successfully built nashpy


In [2]:
import nashpy as nash
import numpy as np



I am going to make a non-zero sum game representing the prisoner's dilema!

In [33]:
A = np.array([[-8,0],[-12,-1]])
B = np.array([[-8,-12],[0,-1]])

rps = nash.Game(A, B)

In [27]:
rps

Bi matrix game with payoff matrices:

Row player:
[[ -8   0]
 [-12  -1]]

Column player:
[[ -8 -12]
 [  0  -1]]

In [28]:
eqs = rps.support_enumeration()

list(eqs)

[(array([1., 0.]), array([1., 0.]))]

In [34]:
iterations = 10

play_counts = rps.fictitious_play(iterations=iterations)

for row_play_count, column_play_count in play_counts:

    print(row_play_count, column_play_count)

[0 0] [0 0]
[0. 1.] [1. 0.]
[1. 1.] [2. 0.]
[2. 1.] [3. 0.]
[3. 1.] [4. 0.]
[4. 1.] [5. 0.]
[5. 1.] [6. 0.]
[6. 1.] [7. 0.]
[7. 1.] [8. 0.]
[8. 1.] [9. 0.]
[9. 1.] [10.  0.]


What if I play the game randomly.  Player 1 only snitches 10% of the time.  But player 2 always snitches

In [30]:
p1 = [.1,.9]
p2 = [1,0]

rps[p1,p2]

array([-11.6,  -0.8])

Player 1 is going to average 11.6 months in the pokey but player 2 less than a month!

In [35]:
play_counts = rps.stochastic_fictitious_play(iterations=iterations)

for row_play_count, column_play_count in play_counts:

    print(row_play_count, column_play_count)

[array([0, 0]), array([0, 0])] (None, None)
[array([1., 0.]), array([1., 0.])] [array([1.00000000e+00, 1.35430458e-11]), array([1.00000000e+00, 1.43246414e-11])]
[array([2., 0.]), array([2., 0.])] [array([1.00000000e+00, 4.11160353e-18]), array([1.00000000e+00, 4.03004298e-18])]
[array([3., 0.]), array([3., 0.])] [array([1.00000000e+00, 4.20662144e-18]), array([1.00000000e+00, 4.28898414e-18])]
[array([4., 0.]), array([4., 0.])] [array([1.00000000e+00, 4.22625066e-18]), array([1.00000000e+00, 4.16238836e-18])]
[array([5., 0.]), array([5., 0.])] [array([1.00000000e+00, 4.42336035e-18]), array([1.00000000e+00, 4.02282297e-18])]
[array([6., 0.]), array([6., 0.])] [array([1.0000000e+00, 4.2097779e-18]), array([1.00000000e+00, 4.08831526e-18])]
[array([7., 0.]), array([7., 0.])] [array([1.00000000e+00, 4.37775149e-18]), array([1.00000000e+00, 4.03611431e-18])]
[array([8., 0.]), array([8., 0.])] [array([1.00000000e+00, 4.27207303e-18]), array([1.00000000e+00, 4.36752076e-18])]
[array([9., 0.

Stochastic should be random but the game is quickly learning an optimal strategy.  I followed NashPy's documentation to do this https://nashpy.readthedocs.io/en/stable/index.html